src/parser.cpp

```diff
@@ -174,61 +168,61 @@ namespace ada::parser {
-                } else {
-                    url.username += unicode::percent_encode(authority_view.substr(0,password_token_location), character_sets::USERINFO_PERCENT_ENCODE);
-                    url.password += unicode::percent_encode(authority_view.substr(password_token_location+1), character_sets::USERINFO_PERCENT_ENCODE);
+                  size_t password_token_location = authority_view.find(':');
+                  password_token_seen = password_token_location != std::string_view::npos;
+
+                  if (!password_token_seen) {
+                    url.username += unicode::percent_encode(authority_view, character_sets::USERINFO_PERCENT_ENCODE);
+                  } else {
+                      url.username += unicode::percent_encode(authority_view.substr(0,password_token_location), character_sets::USERINFO_PERCENT_ENCODE);
+                      url.password += unicode::percent_encode(authority_view.substr(password_token_location+1), character_sets::USERINFO_PERCENT_ENCODE);
+                  }
+                }
+                else {
+                  url.password += unicode::percent_encode(authority_view, character_sets::USERINFO_PERCENT_ENCODE);
                 }
               }
-              else {
-                url.password += unicode::percent_encode(authority_view, character_sets::USERINFO_PERCENT_ENCODE);
-              }
-            }
-            // Otherwise, if one of the following is true:
-            // - c is the EOF code point, U+002F (/), U+003F (?), or U+0023 (#)
-            // - url is special and c is U+005C (\)
-            else if (input_position == input_size || url_data[input_position] == '/' || url_data[input_position] == '?' || (url.is_special() && url_data[input_position] == '\\')) {
-              // If atSignSeen is true and authority_view is the empty string, validation error, return failure.
-              if (at_sign_seen && authority_view.empty()) {
-                url.is_valid = false;
-                return url;
+              // Otherwise, if one of the following is true:
+              // - c is the EOF code point, U+002F (/), U+003F (?), or U+0023 (#)
+              // - url is special and c is U+005C (\)
+              else if (input_position == input_size || url_data[input_position] == '/' || url_data[input_position] == '?' || (url.is_special() && url_data[input_position] == '\\')) {
+                // If atSignSeen is true and authority_view is the empty string, validation error, return failure.
+                if (at_sign_seen && authority_view.empty()) {
+                  url.is_valid = false;
+                  return url;
+                }
+                // Decrease pointer by the number of code points in buffer plus one,
+                // set buffer to the empty string, and set state to host state.
+                input_position -= authority_view.length() + 1;
```

Comment:
We can probably avoid decrementing input_position by reversing the logic: we only increment it in the positive case above (so we increment it conditionally). This would be simpler than incrementing it generally and then decrementing conditionally.